cmd/agent/main.go

```diff
@@ -19,24 +23,43 @@ var (
 
 func main() {
 	flag.Parse()
-	downloader := agent.Downloader{
+	downloader := agent.Downloader {
 		ModelDir:  *modelDir,
-		Providers: map[storage.Protocol]storage.Provider{},
+		Providers: map[kfstorage.Protocol]kfstorage.Provider{},
 	}
-	if endpoint, ok := os.LookupEnv(s3credential.AWSEndpointUrl); ok {
-		region, _ := os.LookupEnv(s3credential.AWSRegion)
-		sess, err := session.NewSession(&aws.Config{
-			Endpoint: aws.String(endpoint),
-			Region:   aws.String(region)},
-		)
+
+	storageUri, err := downloader.GetProvider()
+	if err != nil {
+		panic(err)
+	}
+
+	if strings.Contains(storageUri, string(kfstorage.S3)) {
+		if endpoint, ok := os.LookupEnv(s3credential.AWSEndpointUrl); ok {
+			region, _ := os.LookupEnv(s3credential.AWSRegion)
+			sess, err := session.NewSession(&aws.Config{
+				Endpoint: aws.String(endpoint),
+				Region:   aws.String(region)},
+			)
+			if err != nil {
+				panic(err)
+			}
+			sessionClient := s3.New(sess)
+			downloader.Providers[kfstorage.S3] = &kfstorage.S3Provider{
+				Client: sessionClient,
+				Downloader: s3manager.NewDownloaderWithClient(sessionClient, func(d *s3manager.Downloader) {
+				}),
+			}
+		}
+	}
+
+	if strings.Contains(storageUri, string(kfstorage.GCS)) {
+		ctx := context.Background()
+		client, err := storage.NewClient(ctx)
 		if err != nil {
 			panic(err)
 		}
-		sessionClient := s3.New(sess)
-		downloader.Providers[storage.S3] = &storage.S3Provider{
-			Client: sessionClient,
-			Downloader: s3manager.NewDownloaderWithClient(sessionClient, func(d *s3manager.Downloader) {
-			}),
+		downloader.Providers[kfstorage.GCS] = &kfstorage.GCSProvider{
+			Client: mockapi.AdaptClient(client),
```

Comment:
here should use the real client?

Reply:
The real client is defined by `storage.NewClient(ctx)` on line 54. This line instead adapts the client to fix the mock client interface which is used to allow unit testing. At first I had set the client type for the GCS provider as storage.Client, but this was causing type errors when I went and mocked the client for unit tests - so I used a common interface between the mocked version and the real client. 

I might be missing a better way to do this, so I'll look into it further.